ec2/ec2test/server.go

```diff
@@ -2536,6 +2539,100 @@ func (srv *Server) modifyInstanceAttribute(w http.ResponseWriter, req *http.Requ
 	}
 }
 
+func (srv *Server) createTags(w http.ResponseWriter, req *http.Request, reqId string) interface{} {
+	var resourceIds []string
+	var tags []ec2.Tag
+
+	for attr, vals := range req.Form {
+		if strings.HasPrefix(attr, "ResourceId.") {
+			fields := strings.SplitN(attr, ".", 2)
+			if len(fields) != 2 {
+				fatalf(400, "InvalidParameterValue", "unknown field %s: %s", attr, vals[0])
+			}
+			i := atoi(fields[1])
+			for i > len(resourceIds) {
+				resourceIds = append(resourceIds, "")
+			}
+			resourceIds[i-1] = vals[0]
+		} else if strings.HasPrefix(attr, "Tag.") {
+			fields := strings.SplitN(attr, ".", 3)
+			if len(fields) != 3 {
+				fatalf(400, "InvalidParameterValue", "unknown field %s: %s", attr, vals[0])
+			}
+			i := atoi(fields[1])
+			for i > len(tags) {
+				tags = append(tags, ec2.Tag{})
```

Comment:
> Tag.1.Key
> Tag.1.Value

At this point: tags is empty, `i` is 1. It'll add a single empty ec2.Tag to tags, and then below will set the i-1'th element (i.e. index 0 in this case) to the key/value.

`tags == []ec2.Tag{{"Tag.1.Key", "Tag.1.Value"}}`

> Tag.3.Key
> Tag.3.Value

Now tags has len==1, and `i` is 3. The code will add two empty ec2.Tags to tags, and below will set the element at index 2 to the key/value. So at this point we'd have:

`tags == []ec2.Tag{{"Tag.1.Key", "Tag.1.Value"}, {}, {"Tag.2.Key", "Tag.2.Value"}}`

> Is the above possible? Or is it enforced that there must be a tag 1 and tag 2 and tag 3 etc, not missing values like tag 1 and tag 2 and tag4. Either way, I think another test is needed

The values are going in the right places. I'd say it's an error to have non-contiguous sequence numbers. We don't have to test every possible scenario in the test server, though; only the ones that we actually run. We will never generate such parameters, so it's not worth trying to bend the code to test it.